### ISSUE TITLE
Some podcast episodes aren't saved to database

### ISSUE THREAD

**Issue by neodyne** [gpodder/gpodder#847]
On Windows 10, using gPodder 3.10.16, portable version.

After adding http://www.emptysea.net/tote/podcast1.xml to gPodder (first run, new database), all episodes (currently ~101) are correctly shown in the GUI. However, after restarting gPodder only 22 (in no particular order) are visible in the episodes list. Other feeds can be added and are displayed properly without issues.

The log lists the same error for each missing episode (only the object address changes):

`[gpodder.dbsqlite] ERROR: Cannot save <gpodder.gtkui.model.GEpisode object at 0x04736918>: UNIQUE constraint failed: episode.podcast_id, episode.guid`
`Traceback (most recent call last):`
`  File "C:\apps\gpodder-3.10.16-portable\data\lib\python3.8\site-packages\gpodder\dbsqlite.py", line 214, in _save_object`
`    cur.execute(sql, values)`
`sqlite3.IntegrityError: UNIQUE constraint failed: episode.podcast_id, episode.guid`

**Comment by elelay** [comment]
`<guid>` are missing for all `<item>` except one. I'm surprised you even see more than 2.
gPodder needs guid (they are mandatory in rss) to uniquely identify each episode.
This allows to not mark an episode new even if the title changed for instance.

If possible, reach the publisher of the feed so that they fix the issue.

**Comment by neodyne** [comment]
Thank you for responding.

I don't think missing the (arguably [optional](https://validator.w3.org/feed/docs/rss2.html#ltguidgtSubelementOfLtitemgt)) `<guid>` tag for `<item>` is the issue, I subscribe to many rss feeds without those tags for items, and they parse just fine in gPodder. After looking further into the xml file, it looks like the authors of the feed decided to add an identical `<link>` tag to a few consecutive items in the feed at a time, denoting an individual storyline. The individual episode urls in `<enclosure>`, as well as other relevant item tags, are unique though. For example, the oldest five episodes all have http://emptysea.net/tote/shadow.html in `<link>` (now 404), and unsurprisingly, only the last of the five (inverse order) gets saved to the database, others get discarded as duplicates / throw the error. Looking back, the visible episodes are only the last ones from these storylines.
The xml with the `<link>` tag removed from all `<item>` (still without item guids) parses / saves in the database just fine, because there is no duplicity.

While I understand that this is probably expected behavior in gPodder and a bad decision on the part of the feed's authors, could it be possible to relax the criteria for unique items in the database, so that the an unique url in the `<enclosure>` tag is enough, instead of the `<link>` tag / combination of the two (when the `<guid>` tag is not present)? After all, the iTunes webpage (https://podcasts.apple.com/us/podcast/tales-of-the-extraordinary/id274001489), has no problems listing all items from the same xml; also the feed parses fine in some other podcatchers / programs with podcast functionality that I tried, like Grover, or AIMP, as it seems they do not check for unique `<link>` tags, only `<enclosure>`. Funnily enough, the Spotify webpage for the podcast (https://open.spotify.com/show/15xGuNYs700JdGtoc2iRBE) is broken at the second episode, most likely for the same reason.

Feel free to close the issue, if you feel it is not worth / not reasonable to break spec / convention for this, it just seems it is handled more loosely in other programs.

**Comment by elelay** [comment]
Ah yes, you're totally right: [if no guid it falls back to link. If no link it uses the enclosure.](https://github.com/gpodder/podcastparser/blob/master/podcastparser.py#L728). Bad thing in this feed is they put thee same link in many episodes.

I don't intend to relax this constraint so I'm closing the issue. Thanks again for the clear write-up.

**Comment by auouymous** [comment]
The enclosure can change if the publisher switches CDN's or adds/changes a download tracker. I've seen feeds that do this often.

The pubDate field converted to a unix timestamp would be more stable (assuming publisher doesn't change it), but it could be possible for multiple items to exist with the same pubDate (could also have multiple items with same title). Also, this particular feed uses invalid pubDate fields, maybe they could add guids while fixing the dates.

**Comment by auouymous** [comment]
Although, the enclosure would be unique, even if not stable.

I do believe something should be done about this issue. Gpodder should either fallback to enclosure or detect the missing GUID and properly handle it. Right now it displays the episodes that lack GUIDs and allows them to be downloaded, but it fails to save those episodes to the database. At next restart, you no longer see the episodes and get unknown external files for any you downloaded. A broken feed shouldn't create broken behavior in gpodder.

dbsqlite.py `_save_object()` can store an error message in `o.parent._update_error` which is visible on the channel entry. It can also do `o.is_new = False` to prevent automatic download of episodes that failed to save. Maybe also mark them as deleted with `o.state = gpodder.STATE_DELETED`. The only problem is that they still show in the download window and the user can click their checkboxes to download.

1) It might be best if gpodder never creates objects for episodes that lack a GUID. And write an error to `<channel>._update_error` so the user knows the feed is broken and some/all episodes were removed. The message would also tell them to contact publisher to fix the issue.

2) It could probably also fallback to enclosure but do something like `<channel>._update_error = 'feed uses unstable GUIDs for episodes, may result in duplicate episodes...'`. This message would also tell them contact publisher to fix it. This is better for users but optional 1 is more likely to result in the publisher fixing the problem when users complain.

**Comment by neodyne** [comment]
In my opinion, relying solely on guid (1) as a mandatory element would overstep the RSS 2.0 Specification - technically the only required fields for rss items are at least title or description, nothing more. If enclosure is present, the explicit media length (0 as a minimum) and type (and url of course) is mandatory. In general, guid-less feeds are not broken according to spec (and gPodder correctly does not treat them as such currently). Furthermore, I have seen many feeds that use the url for the shownotes page for the episode as link, even guid, since it's parsed as string anyways. With multipart episodes (with identical shownotes), I am quite surprised this did not happen sooner.

In regards to (2), I see picking the element you base "the uniqueness" of a feed item on (in the absence of guid) as a trade-off. As you pointed out, basing it on title, pubDate, and now as we see, the link element each can produce edge cases. Personally, I haven't seen a feed with 2 consecutive items having the same enclosure url, but have seen feeds with episodes dumped at the same time (same timestamp), or mislabeled with the same title (albeit most likely in error). Maybe a different fallback sequence, e.g. guid > enclosure > title > link, would be worth considering.

Although I agree that the behavior when adding the "broken" feed is inconsistent. I usually only stream podcasts, but until gPodder is restarted, nothing indicates a problem, all episodes can be checked to be added, then streamed without problem. I can imagine it's more obvious with downloaded episodes, and therefore should be properly indicated somehow.

**Comment by auouymous** [comment]
Gpodder doesn't create an episode if there is no enclosure. Therefore, the sequence should be `guid > enclosure`, because no enclosure means no episode. Unless #849 is fully implemented, then additional fallbacks such as title and link would need to be considered.

The problem is that non-unique GUIDs can't be saved to the database. A duplicate enclosure, title or link results in the episode not be saved at all, or updates to an existing episode not being saved. So gpodder must discard the episode if it detects a collision with another. Falling back to enclosure would mostly eliminate this issue for media feeds since most episodes would have unique enclosures.

A full implementation of #849 would need additional fallbacks to title and link, or simply discard episodes without guid tags. Proper collision handling and error reporting needed for this issue would inform users of broken news/blog feeds. Not much else could be done for those types of feeds.

**Comment by elelay** [comment]
A long time ago, gPodder used [feedparser](https://pythonhosted.org/feedparser/) to try very hard to parse any feed on earth. This resulted in more and more workarounds and a slow program.
Now we try a stricter approach, where we encourage users to report back broken feeds. They get fixed more often than not.

So I'm all for adding an explicit error message but not for additional workarounds.

**Comment by auouymous** [comment]
The fallback happens at https://github.com/gpodder/podcastparser/blob/master/podcastparser.py#L728 and uses `guid > link > enclosure`. Swapping link and enclosure wouldn't add any additional workarounds, and enclosure is likely to be more unique than link. For media podcasts, the enclosure is everything and `link > enclosure` allows the same file to be downloaded for multiple episodes. Swapping the order drops older episodes and only downloads the file once.

Podcastparser currently discards the episode if it has multiple enclosures and no link. Swapping order would fallback to link if multiple enclosures and discard if not found. Adding an additional fallback to title would only slow down feeds that meet this pattern.

When swapping the fallback order, all but one episode in this issue's podcast work. Two of them have different titles but the same link and enclosure. #850 deletes episode 80 and keeps the newer episode 81.

I have a working patch for podcastparser to swap order, should I post it? If so, should I include a title fallback after link? But the only way to inform the user is by having podcastparser add an additional field to the episode entry, toggle it when using a guid fallback and have gpodder check that field. Should I do that or leave it as is and not inform the user?
